Add ability to mount volumes and specify environment variables when using mlflow with docker

nlml opened the pull request:
## What changes are proposed in this pull request?

In addition to the `image` field, I've added a `volumes` field when specifying a docker environment through the `MLproject` file. This allows users to mount local volumes when using `mlflow run ...` with docker. I need this so my data is accessible to the training run. I also specified an `environment` field that can also be added to `MLproject`; here users can specify environment variables to be passed to the `docker run` command by mlflow.

## How is this patch tested?

I am currently using this to run experiments and it is all working as expected. I can write unit tests if desired, but these changes will have no effect unless users specify either the `volumes` or `environment` field under the docker section of their `MLproject` yaml config.

## Release Notes

Now able to specify volumes to mount when using mlflow with docker image.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Yes: you can add another field below `image` called `volumes` in your MLproject.yaml. It should be a list of strings e.g. ["/path/to/mount:/path/to/mount", "/another/path:/another/path"]

### What component(s) does this PR affect?

- [ ] UI
- [ ] CLI
- [ ] API
- [ ] REST-API
- [ ] Examples
- [ ] Docs
- [ ] Tracking
- [x] Projects
- [ ] Artifacts
- [ ] Models
- [ ] Scoring
- [ ] Serving
- [ ] R
- [ ] Java
- [ ] Python

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

nlml committed:
add support for extra docker volumes

nlml committed:
assert any provided docker volumes are in valid format

Commit:
add ability to specify environment variables in MLproject file when using Docker

Commit:
fix linter error

tallen94 commented:
Hello! I was working on adding a similar feature! In my case it was useful to forward on the system environment variables to the docker container because they were sensitive credentials. In this case, if an MLProject were checked into source, it would be contained in there. What do you think of modifying this such that it also passes in variables that are available from the system itself?

Here is the branch that I created: https://github.com/tallen94/mlflow/tree/user-envs

In this case the yml file would look something like:
```
docker_env:
  image:  mlflow-docker-test
  envs:
    - DB_USER
    - DB_PASS
```

nlml commented:
Yep great idea! I'll add this functionality to my PR. Thanks!

smurching reviewed (commented):
@nlml thanks for the PR! It would be good to add unit tests - I was also wondering a few things:

* What happens when the volumes or environment variables in the MLproject file are unavailable? Will project execution fail?
* How does this work with the [Kubernetes remote project execution mode](https://www.mlflow.org/docs/latest/projects.html#kubernetes-execution)? I guess currently we won't include the environment variables or volumes while submitting project execution to Kubernetes?

nlml commented:
@smurching no worries!

I will get around to the unit tests later this week I hope.

Project execution will not fail when volumes or environment variables are not specified in the MLproject file. It will just execute as per before my pull request (so my changes will have no effect).

I don't have any familiarity with Kubernetes mode. But I guess this should also have no effect there as AFAIK there is no `docker_env` top-level key in the `MLproject` file when using Kubernetes.

Commit:
fix merge conflicts

Commit:
Merge branch 'mlflow-master'

Commit:
add ability to copy env vars from host system in docker mode

Commit:
add test_docker_user_specified_env_vars

Commit:
fix for copying env vars to docker

nlml commented:
Alright @smurching, I've added unit tests and incorporated the suggestion of @tallen94 into my fork, plus merged master into my fork. So once the CI passes everything looks okay on my end, but let me know if you need something else here :)

nlml commented:
...I'm not sure why the linter failed here?

smurching commented on tests/projects/test_docker_projects.py:
```diff
@@ -242,3 +242,46 @@ def test_docker_unknown_uri_artifact_cmd_and_envs():
+
+    if "should_crash" in expected:
+        expected.remove("should_crash")
+        with pytest.raises(KeyError) as excinfo:
```
```suggestion
        with pytest.raises(KeyError):
```

To fix the lint error, which you can see in https://travis-ci.org/mlflow/mlflow/jobs/608360341?utm_medium=notification&utm_source=github_status

smurching commented:
@nlml thanks for the updates, sorry I should have noticed/mentioned this last time - could we also update the docs at https://github.com/mlflow/mlflow/blob/master/docs/source/projects.rst#specifying-an-environment to describe the new `environment` and `volume` fields?

smurching commented on mlflow/projects/__init__.py:
```diff
@@ -651,6 +653,18 @@ def _get_docker_command(image, active_run):
+                env_vars[user_entry[0]] = user_entry[1]
+            else:
+                # User wants to copy an environment variable from system environment
+                env_vars[user_entry] = os.environ[user_entry]
```
```suggestion
                system_var = os.environ.get(user_entry)
                if system_var is None:
                    raise MlflowException("This project expects the %s environment variables to be set on the machine running the project, but %s was not set. Please ensure all expected environment variables are set" % (", ".join(user_env_vars), user_entry))
```

The idea is basically to throw an informative error message telling the user which environment variables must be set in order to run the project

smurching commented on mlflow/projects/_project_spec.py:
```diff
@@ -36,9 +36,24 @@ def load_project(directory):
-                                 "attribute found.")
+    if docker_env:
+        if not docker_env.get("image"):
+            raise ExecutionException("Docker environment specified but no image "
```
```suggestion
            raise ExecutionException("Project configuration (MLproject file) was invalid: docker environment specified but no image "
```

smurching commented on mlflow/projects/_project_spec.py:
```diff
@@ -36,9 +36,24 @@ def load_project(directory):
+        if docker_env.get("volumes"):
+            if not (isinstance(docker_env["volumes"], list)
+                    and all([isinstance(i, str) for i in docker_env["volumes"]])):
+                raise ExecutionException("Docker volumes must be a list of strings, "
```
```suggestion
                raise ExecutionException("Project configuration (MLproject file) was invalid: Docker volumes must be a list of strings, "
```

smurching reviewed (commented):
Mostly looks good, just some small comments/suggestions - thanks! BTW, you should be able to reproduce the lint failures locally / make sure lint passes by running `./lint.sh` from your checkout of MLflow :)

nlml committed:
Confirm user-requested env var actually exists (docker)

Co-Authored-By: Siddharth Murching <smurching@gmail.com>

Commit:
improve error desc strings

nlml committed:
Catch exception as excinfo in test

Co-Authored-By: Siddharth Murching <smurching@gmail.com>

Commit:
fix lint error

Commit:
fix missing line and docker tests

Commit:
update docs

Commit:
improve docs for docker

nlml commented:
Ok I've addressed all your changes and added docs. Hopefully good to go now :)

smurching commented on mlflow/projects/__init__.py:
```diff
@@ -651,6 +653,25 @@ def _get_docker_command(image, active_run):
+                system_var = os.environ.get(user_entry)
+                if system_var is None:
+                    raise MlflowException(
+                        "This project expects the %s environment variable to "
```
```suggestion
                        "This project expects the %s environment variables to "
```

smurching reviewed (approved):
Thanks for the updates! I had a one-character nit on one of the error messages - was hoping to commit it to your branch myself, but looks like I don't have permission. I'm happy to merge as is if you're able to make a follow-up PR to update the error message wording :)

nlml commented:
Ok great, thanks for merging - added a PR for the typo here: https://github.com/mlflow/mlflow/pull/2074
